cluster/src/main/scala/org/apache/spark/util/DynamicURLClassLoader.scala

```diff
@@ -0,0 +1,156 @@
+      url.getFile.substring(0, url.getFile.lastIndexOf(JOB_SERVER_TEXT))
+    } else url.getFile
+
+    val loader = getClassLoader(url)
+    def oldURL = if (urlList.containsKey(key)) urlList.get(key).getURLs.head.getPath else ""
+
+    if (!oldURL.isEmpty && !oldURL.equals(url.getPath)) {
+      // try to remove the old file. it is  okay if not successful
+      try {
+        logInfo(" Deleting the Loader with key:" + key + " with File: " + oldURL )
+        new File(oldURL).delete()
+      } catch {
+        case _: Throwable => {
+          // do nothing.
+        }
+      }
+    }
+
+    logInfo(" Adding new Loader with key:" + key + " with File: " + url.getPath )
+    urlList.put(key, loader)
+  }
+
+  private def getClassLoader(url: URL): MutableURLClassLoader = {
+    if (parentFirst) new MutableURLClassLoader(Array(url), parent)
+    else new ChildFirstURLClassLoader(Array(url), parent)
+  }
+
+  override def getURLs(): Array[URL] = {
+    urlList.asScala.flatMap(_._2.getURLs).toArray ++ parent.getURLs
+  }
+
+  override def getResources(name: String): util.Enumeration[URL] = {
+    val childResources =
+      urlList.asScala.map(tuple => {
+        tuple._2.getResources(name).asScala
+            .filter(!parent.getResources(name).asScala.contains(_)).asJavaEnumeration
+      }).toArray
+
+    val parentResources = Array(parent.getResources(name).asInstanceOf[util.Enumeration[_]])
+
+    new CompoundEnumeration(childResources ++ parentResources)
+  }
+
+  @throws[ClassNotFoundException]
+  override protected def loadClass(className: String, resolve: Boolean): Class[_] = {
+    var c: Class[_] = findLoadedClass(className)
+
+    if (c != null) {
+      if (resolve) resolveClass(c)
+      return c
+    }
+
+    if (parentFirst) {
+      c = loadFromParent(parent, className, false)
+          .getOrElse(loadFromJars(className, false)
+              .getOrElse(loadFromStore(className, true).get))
+    } else {
+      c = loadFromJars(className, false)
+          .getOrElse(loadFromStore(className, false)
+              .getOrElse(loadFromParent(parent, className, true).get))
+    }
+
+    if (resolve) {
+      resolveClass(c)
+    }
+
+    c
+  }
+
+  private def loadClassFunction(f: () => Option[Class[_]], throwException: Boolean = false)
+  : Option[Class[_]] = {
+    var clazz: Option[Class[_]] = None
+    try {
+      f()
+    } catch {
+      case cnfe: ClassNotFoundException => if (throwException) throw cnfe else None
+    }
+  }
+
+  @throws[ClassNotFoundException]
+  private def loadFromJars(className: String, throwException: Boolean): Option[Class[_]] = {
+    var c: Option[Class[_]] = None
+    for (url <- urlList.asScala if (!c.isDefined)) {
+      c = loadClassFunction(() => Some(url._2.loadClass(className)), false)
+    }
+
+    c match {
+      case Some(_) => c
+      case None =>
+        if (throwException) throw new ClassNotFoundException(className)
+        else None
+    }
+
+  }
```

Comment:
Don't understand the purpose of this. Is this something to do with how job-server is loading and does it have its own classloader that this is trying to emulate (and add to)?